{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import gffutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_contigs_df = pd.read_csv('./CM3/final_results/abyss/A_contigs.tsv', sep='\\t')\n",
    "A_contigs_df['contig_length'] = A_contigs_df['contig'].apply(lambda x: len(x))\n",
    "A_contigs_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "db_path = 'd:/scRNA/Thesis-Code-Base-master/Data/annotation/'\n",
    "psl_file = './CM3/final_results/abyss/A_contigs_alignment.psl'\n",
    "contigs_file ='./CM3/final_results/abyss/A_contigs.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = gffutils.FeatureDB(db_path + \"GENCODE_v45.db\")\n",
    "polyA_db = gffutils.FeatureDB(db_path + \"polyA_v45.db\")\n",
    "tRNA_db = gffutils.FeatureDB(db_path + \"tRNA_v45.db\")\n",
    "psl_df = pd.read_csv(psl_file, delimiter='\\t', comment='#', header=None, names=['name', 'subject', 'identity', 'length', 'mismatches', 'gap_openings', 'q_start', 'q_end', 's_start', 's_end', 'e_value', 'bit_score'])\n",
    "contigs_df = pd.read_csv(contigs_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "contigs_df['contig_length'] = contigs_df['contig'].apply(lambda x: len(x))\n",
    "contigs_df = contigs_df[contigs_df['contig_length'] >= 31]\n",
    "psl_df = psl_df.merge(contigs_df, on='name', how='left')\n",
    "psl_df = psl_df[psl_df['mismatches']/psl_df['contig_length']<=0.05]\n",
    "psl_df = psl_df[psl_df['contig_length']==psl_df['length']]\n",
    "psl_df.sort_values(by='e_value', inplace=True)\n",
    "print('psl count: ', psl_df.shape[0] )\n",
    "lncRNAs = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "pseudoGenes = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "retained_introns = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "transcribed_unprocessed_pseudogenes = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "miRNAs = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "polyAs = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "tRNAs = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "exons = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "unmapped = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "CDS = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "UTR = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "\n",
    "gene = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "\n",
    "start_codon = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "\n",
    "stop_codon = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}\n",
    "\n",
    "transcript = {\n",
    "    'name': [],\n",
    "    'contig': [],\n",
    "    'chr': [],\n",
    "    'start': [],\n",
    "    'end': [],\n",
    "    'gene_name': [],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_to_annotation_dict(annotation_dict, feature, name, contig, chr, start, end):\n",
    "    annotation_dict['name'].append(name)\n",
    "    annotation_dict['contig'].append(contig)\n",
    "    annotation_dict['chr'].append(chr)\n",
    "    annotation_dict['start'].append(start)\n",
    "    annotation_dict['end'].append(end)\n",
    "    annotation_dict['gene_name'].append(feature.attributes['gene_name'][0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt = 0\n",
    "f_types = set()\n",
    "for index, row in psl_df.iterrows():\n",
    "    chr, start, end = row['subject'], row['s_start'], row['s_end']\n",
    "    features = db.region(region=(chr, start, end))\n",
    "    for f in features:\n",
    "        f_types.add(f.featuretype)\n",
    "        if f.attributes['gene_type'][0]=='lncRNA':\n",
    "            add_to_annotation_dict(lncRNAs, f, row['name'], row['contig'], chr, start, end)\n",
    "        if f.attributes['gene_type'][0]=='miRNA':\n",
    "            add_to_annotation_dict(miRNAs, f, row['name'], row['contig'], chr, start, end)\n",
    "        if 'transcript_type' in f.attributes:\n",
    "            if f.attributes['transcript_type'][0]=='retained_intron':\n",
    "                add_to_annotation_dict(retained_introns, f, row['name'], row['contig'], chr, start, end)\n",
    "        if f.attributes['gene_type'][0]=='processed_pseudogene':\n",
    "            add_to_annotation_dict(pseudoGenes, f, row['name'], row['contig'], chr, start, end)\n",
    "        if f.attributes['gene_type'][0]=='transcribed_processed_pseudogene':\n",
    "            add_to_annotation_dict(transcribed_unprocessed_pseudogenes, f, row['name'], row['contig'], chr, start, end)\n",
    "        if f.featuretype=='exon':\n",
    "            add_to_annotation_dict(exons, f, row['name'], row['contig'], chr, start, end)\n",
    "        if f.featuretype=='CDS':\n",
    "            add_to_annotation_dict(CDS, f, row['name'], row['contig'], chr, start, end)\n",
    "        if f.featuretype=='UTR':\n",
    "            add_to_annotation_dict(UTR, f, row['name'], row['contig'], chr, start, end)\n",
    "        if f.featuretype=='gene':\n",
    "            add_to_annotation_dict(gene, f, row['name'], row['contig'], chr, start, end)\n",
    "        if f.featuretype=='start_codon':\n",
    "            add_to_annotation_dict(start_codon, f, row['name'], row['contig'], chr, start, end)\n",
    "        if f.featuretype=='stop_codon':\n",
    "            add_to_annotation_dict(stop_codon, f, row['name'], row['contig'], chr, start, end)\n",
    "        if f.featuretype=='transcript':\n",
    "            add_to_annotation_dict(transcript, f, row['name'], row['contig'], chr, start, end)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in psl_df.iterrows():\n",
    "    chr, start, end = row['subject'], row['s_start'], row['s_end']\n",
    "    features = polyA_db.region(region=(chr, start, end))\n",
    "    for f in features:\n",
    "        add_to_annotation_dict(polyAs, f, row['name'], row['contig'], chr, start, end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in psl_df.iterrows():\n",
    "  chr, start, end = row['subject'], row['s_start'], row['s_end']\n",
    "  features = tRNA_db.region(region=(chr, start, end))\n",
    "  for f in features:\n",
    "      add_to_annotation_dict(tRNAs, f, row['name'], row['contig'], chr, start, end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "lncRNAs_df = pd.DataFrame(lncRNAs)\n",
    "lncRNAs_df.to_csv(contigs_file.replace('.tsv', '_lncRNA.tsv'), sep='\\t', index=False) \n",
    "\n",
    "retained_introns_df = pd.DataFrame(retained_introns)\n",
    "retained_introns_df.to_csv(contigs_file.replace('.tsv', '_retained_introns.tsv'), sep='\\t', index=False)\n",
    "\n",
    "pseudoGenes_df = pd.DataFrame(pseudoGenes)\n",
    "pseudoGenes_df.to_csv(contigs_file.replace('.tsv', '_pseudoGenes.tsv'), sep='\\t', index=False)\n",
    "\n",
    "transcribed_unprocessed_pseudogenes_df = pd.DataFrame(transcribed_unprocessed_pseudogenes)\n",
    "transcribed_unprocessed_pseudogenes_df.to_csv(contigs_file.replace('.tsv', '_transcribed_unprocessed_pseudogenes.tsv'), sep='\\t', index=False)\n",
    "\n",
    "polyAs_df = pd.DataFrame(polyAs)\n",
    "polyAs_df.to_csv(contigs_file.replace('.tsv', '_polyAs.tsv'), sep='\\t', index=False)\n",
    "\n",
    "tRNAs_df = pd.DataFrame(tRNAs)\n",
    "tRNAs_df.to_csv(contigs_file.replace('.tsv', '_tRNAs.tsv'), sep='\\t', index=False)\n",
    "\n",
    "miRNAs_df = pd.DataFrame(miRNAs)\n",
    "miRNAs_df.to_csv(contigs_file.replace('.tsv', '_miRNAs.tsv'), sep='\\t', index=False)\n",
    "exons_df = pd.DataFrame(exons)  \n",
    "\n",
    "exons_df.to_csv(contigs_file.replace('.tsv', '_exons.tsv'), sep='\\t', index=False)\n",
    "CDS_df = pd.DataFrame(CDS)\n",
    "CDS_df.to_csv(contigs_file.replace('.tsv', '_CDS.tsv'), sep='\\t', index=False)\n",
    "\n",
    "UTR_df = pd.DataFrame(UTR)\n",
    "UTR_df.to_csv(contigs_file.replace('.tsv', '_UTR.tsv'), sep='\\t', index=False)\n",
    "start_codon_df = pd.DataFrame(start_codon)\n",
    "start_codon_df.to_csv(contigs_file.replace('.tsv', '_start_codon.tsv'), sep='\\t', index=False)\n",
    "stop_codon_df = pd.DataFrame(stop_codon)\n",
    "stop_codon_df.to_csv(contigs_file.replace('.tsv', '_stop_codon.tsv'), sep='\\t', index=False)\n",
    "gene_df = pd.DataFrame(gene)\n",
    "gene_df.to_csv(contigs_file.replace('.tsv', '_gene.tsv'), sep='\\t', index=False)\n",
    "transcript_df = pd.DataFrame(transcript)\n",
    "transcript_df.to_csv(contigs_file.replace('.tsv', '_transcript.tsv'), sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('lncRNAs count: ', lncRNAs_df.shape[0])\n",
    "print('retained_introns count: ', retained_introns_df.shape[0])\n",
    "print('pseudoGenes count: ', pseudoGenes_df.shape[0])\n",
    "print('transcribed_unprocessed_pseudogenes count: ', transcribed_unprocessed_pseudogenes_df.shape[0])\n",
    "print('polyAs count: ', polyAs_df.shape[0])\n",
    "print('tRNAs count: ', tRNAs_df.shape[0])\n",
    "print('miRNAs count: ', miRNAs_df.shape[0])\n",
    "print('exons count: ', exons_df.shape[0])\n",
    "print('CDS count: ', CDS_df.shape[0])\n",
    "print('UTR count: ', UTR_df.shape[0])\n",
    "print('start_codon count: ', start_codon_df.shape[0])\n",
    "print('stop_codon count: ', stop_codon_df.shape[0])\n",
    "print('gene count: ', gene_df.shape[0])\n",
    "print('transcript count: ', transcript_df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_mapped_contigs = pd.concat([lncRNAs_df, \n",
    "                                retained_introns_df, \n",
    "                                pseudoGenes_df, \n",
    "                                transcribed_unprocessed_pseudogenes_df, \n",
    "                                polyAs_df, \n",
    "                                tRNAs_df, \n",
    "                                miRNAs_df,\n",
    "                                exons_df,\n",
    "                                CDS_df,\n",
    "                                UTR_df,\n",
    "                                start_codon_df,\n",
    "                                stop_codon_df,\n",
    "                                gene_df,\n",
    "                                transcript_df\n",
    "                                ])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unannotated_df = psl_df[~psl_df['name'].isin(all_mapped_contigs['name'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unannotated_df.to_csv(contigs_file.replace('.tsv', '_unannotated.tsv'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmapped_df = contigs_df[~contigs_df['name'].isin(psl_df['name'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmapped_df.to_csv(contigs_file.replace('.tsv', '_unmapped.tsv'), sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
