{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_kmers_df = pd.read_csv('filtered_kmers.tsv', sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_kmers_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers = all_kmers_df[(all_kmers_df['log2FoldChange'] > 1) & (all_kmers_df['padj'] < 0.05)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_contigs = pd.read_csv('A_contigs.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each contig, find the constituent kmers of length 31 and store them in a column of the df as a list\n",
    "A_contigs['kmers'] = A_contigs['contig'].apply(lambda x: [x[i:i+31] for i in range(len(x)-31)])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_contigs['kmer_count'] = A_contigs['kmers'].apply(lambda x: len(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_contigs = A_contigs[A_contigs['kmer_count'] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each contig, take the kmers and find their padj values and store in another column\n",
    "A_contigs['padj'] = A_contigs['kmers'].apply(lambda x: [kmers.loc[kmer, 'padj'] for kmer in x if kmer in kmers.index])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_annotations = pd.read_csv('A_contigs_annotation.tsv', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_intron_annotation = pd.read_csv('A_contigs_annotation_intron.tsv', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_annotations = pd.concat([A_annotations, A_intron_annotation])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_annotations = A_annotations.drop_duplicates(subset=['contig', 'chromosome', 'start', 'end', 'feature_type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_unmapped = A_annotations[A_annotations['feature_type'] == '*']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_intron = A_annotations[(A_annotations['feature_type'] == 'intron')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_CDS = A_annotations[(A_annotations['feature_type'] == 'CDS')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_Junction = A_annotations[(A_annotations['feature_type'] == 'exon')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_homology = pd.read_csv('A_contigs_homology_annotation.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_rRNAs = A_homology[A_homology['annotation'] == 'rRNA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_unmapped = A_unmapped.drop_duplicates(subset=['contig'])\n",
    "A_intron   = A_intron.drop_duplicates(subset=['contig'])\n",
    "A_CDS      = A_CDS.drop_duplicates(subset=['contig'])\n",
    "A_Junction = A_Junction.drop_duplicates(subset=['contig'])\n",
    "A_rRNA     = A_rRNAs.drop_duplicates(subset=['contig'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_unmapped_kmers = []\n",
    "A_intron_kmers = []\n",
    "A_CDS_kmers = []\n",
    "A_Junction_kmers = []\n",
    "A_rRNA_kmers = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_unmapped_contigs = A_contigs[A_contigs['contig'].isin(A_unmapped['contig'].to_list())]\n",
    "A_intron_contigs   = A_contigs[A_contigs['contig'].isin(A_intron['contig'].to_list())]\n",
    "A_CDS_contigs      = A_contigs[A_contigs['contig'].isin(A_CDS['contig'].to_list())]\n",
    "A_Junction_contigs = A_contigs[A_contigs['contig'].isin(A_Junction['contig'].to_list())]\n",
    "A_rRNA_contigs     = A_contigs[A_contigs['contig'].isin(A_rRNA['contig'].to_list())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx, row in A_unmapped_contigs.iterrows():\n",
    "    A_unmapped_kmers.extend(row['kmers'])\n",
    "\n",
    "for idx, row in A_intron_contigs.iterrows():\n",
    "    A_intron_kmers.extend(row['kmers'])\n",
    "\n",
    "for idx, row in A_CDS_contigs.iterrows():\n",
    "    A_CDS_kmers.extend(row['kmers'])\n",
    "\n",
    "for idx, row in A_Junction_contigs.iterrows():\n",
    "    A_Junction_kmers.extend(row['kmers'])\n",
    "\n",
    "for idx, row in A_rRNA_contigs.iterrows():\n",
    "    A_rRNA_kmers.extend(row['kmers'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_unmapped_kmers = list(set(A_unmapped_kmers))\n",
    "A_intron_kmers   = list(set(A_intron_kmers))\n",
    "A_CDS_kmers      = list(set(A_CDS_kmers))\n",
    "A_Junction_kmers = list(set(A_Junction_kmers))\n",
    "A_rRNA_kmers     = list(set(A_rRNA_kmers))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_unmapped_kmers_padj = [kmers.loc[kmer, 'padj'] for kmer in A_unmapped_kmers if kmer in kmers.index]\n",
    "A_intron_kmers_padj   = [kmers.loc[kmer, 'padj'] for kmer in A_intron_kmers if kmer in kmers.index]\n",
    "A_CDS_kmers_padj      = [kmers.loc[kmer, 'padj'] for kmer in A_CDS_kmers if kmer in kmers.index]\n",
    "A_Junction_kmers_padj = [kmers.loc[kmer, 'padj'] for kmer in A_Junction_kmers if kmer in kmers.index]\n",
    "A_rRNA_kmers_padj     = [kmers.loc[kmer, 'padj'] for kmer in A_rRNA_kmers if kmer in kmers.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# normalize the padj values\n",
    "# take the negative log of the padj values\n",
    "import numpy as np\n",
    "A_unmapped_kmers_padj_norm = [-np.log10(x) for x in A_unmapped_kmers_padj]\n",
    "A_intron_kmers_padj_norm   = [-np.log10(x) for x in A_intron_kmers_padj]\n",
    "A_CDS_kmers_padj_norm      = [-np.log10(x) for x in A_CDS_kmers_padj]\n",
    "A_Junction_kmers_padj_norm = [-np.log10(x) for x in A_Junction_kmers_padj]\n",
    "A_rRNA_kmers_padj_norm     = [-np.log10(x) for x in A_rRNA_kmers_padj]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "import plotly.io as pio\n",
    "\n",
    "data = [A_Junction_kmers_padj_norm,\n",
    "        A_CDS_kmers_padj_norm,\n",
    "        A_intron_kmers_padj_norm,\n",
    "        A_unmapped_kmers_padj_norm,\n",
    "        A_rRNA_kmers_padj_norm\n",
    "        ]\n",
    "\n",
    "# Labels and colors for each category\n",
    "labels = ['Junction', 'CDS', 'Intron', 'Unmapped', 'rRNA']\n",
    "# Create the boxplot with customized layout\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add a box for each category with a unique color\n",
    "for d, label in zip(data, labels):\n",
    "    fig.add_trace(go.Box(\n",
    "        y=d,\n",
    "        name=label\n",
    "    ))\n",
    "\n",
    "# Limit y-axis range\n",
    "fig.update_yaxes(range=[0, 15])\n",
    "\n",
    "# Customize layout to show axis lines\n",
    "fig.update_layout(\n",
    "    yaxis_title='-log10(padj)',\n",
    "    xaxis_title='Category',\n",
    "    plot_bgcolor='white',\n",
    "    paper_bgcolor='white',\n",
    "    font=dict(size=14),\n",
    "    xaxis=dict(showgrid=True, zeroline=False, showline=True, linecolor='black', linewidth=2),\n",
    "    yaxis=dict(showgrid=True, zeroline=False, showline=True, linecolor='black', linewidth=2),\n",
    ")\n",
    "\n",
    "fig.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
